### PR TITLE
ICU-21810 Fix error checking of ucurr.cpp part 1

### PR DESCRIPTION
**Pull request by FrankYFTang**
Change code inside ucurr.cpp to use LocalPointer  to prepare the later changes for correct memory change

##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21810
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/common/ucurr.cpp [hunk 1]
```diff
@@ -954,14 +938,18 @@ toUpperCase(const char16_t* source, int32_t len, const char* locale) {
 
     ec = U_ZERO_ERROR;
     dest = (char16_t*)uprv_malloc(sizeof(char16_t) * MAX(destLen, len));
+    if (dest == nullptr) {
+        return dest;
```
[the remark this answers is not in the document]
Comment (FrankYFTang) [reply icu4c/source/common/ucurr.cpp:942]: set U_MEMORY_ALLOCATION_ERROR to what? nothing else return here. ec is local variable before/after my change here.

File: icu4c/source/common/ucurr.cpp [hunk 2]
```diff
@@ -1009,113 +1002,86 @@ collectCurrencyNames(const char* locale,
...
             // Add currency long name.
-            s = ures_getStringByIndex(names, UCURR_LONG_NAME, &len, &ec2);
-            (*currencyNames)[*total_currency_name_count].IsoCode = iso;
+            s = ures_getStringByIndex(names.getAlias(), UCURR_LONG_NAME, &len, &ec2);
             char16_t* upperName = toUpperCase(s, len, locale);
-            (*currencyNames)[*total_currency_name_count].currencyName = upperName;
-            (*currencyNames)[*total_currency_name_count].flag = NEED_TO_BE_DELETED;
-            (*currencyNames)[(*total_currency_name_count)++].currencyNameLen = len;
+
+            if (upperName != nullptr) {
```
[the remark this answers is not in the document]
Comment (FrankYFTang) [reply icu4c/source/common/ucurr.cpp:1041]: ok, I see. my origional intend is to fix problem, not to migrate the code. But I can split that into two PR if you think that is better.

File: icu4c/source/common/ucurr.cpp [hunk 3]
```diff
@@ -1486,15 +1458,19 @@ getCacheEntry(const char* locale, UErrorCode& ec) {
...
-            uprv_strcpy(cacheEntry->locale, locale);
-            cacheEntry->currencyNames = currencyNames;
-            cacheEntry->totalCurrencyNameCount = total_currency_name_count;
-            cacheEntry->currencySymbols = currencySymbols;
-            cacheEntry->totalCurrencySymbolCount = total_currency_symbol_count;
-            cacheEntry->refCount = 2; // one for cache, one for reference
-            currentCacheEntryIndex = (currentCacheEntryIndex + 1) % CURRENCY_NAME_CACHE_NUM;
-            ucln_common_registerCleanup(UCLN_COMMON_CURRENCY, currency_cleanup);
+            if (cacheEntry != nullptr) {
+                currCache[currentCacheEntryIndex] = cacheEntry;
```
[the remark this answers is not in the document]
Comment (FrankYFTang) [reply icu4c/source/common/ucurr.cpp:1462]: ok

File: icu4c/source/common/ucurr.cpp [hunk 4]
```diff
@@ -2291,27 +2258,23 @@ ucurr_openISOCurrencies(uint32_t currType, UErrorCode *pErrorCode) {
 U_CAPI int32_t U_EXPORT2
 ucurr_countCurrencies(const char* locale, 
                  UDate date, 
-                 UErrorCode* ec)
-{
+                 UErrorCode* ec) {
```
[the remark this answers is not in the document]
Comment (FrankYFTang) [reply icu4c/source/common/ucurr.cpp:2261]: The current format make it hard to read now. Would you accept a PR only change the format later?

File: icu4c/source/common/ucurr.cpp [hunk 5]
```diff
@@ -976,28 +964,33 @@ collectCurrencyNames(const char* locale,
...
+        return;
 
-    if(currencyNames == nullptr || currencySymbols == nullptr) {
-      ec = U_MEMORY_ALLOCATION_ERROR;
     }
 
-    if (U_FAILURE(ec)) return;
+    *currencySymbols = (CurrencyNameStruct*)uprv_malloc(sizeof(CurrencyNameStruct) * (*total_currency_symbol_count));
+    if (*currencySymbols == nullptr) {
+        uprv_free(*currencyNames);
```
[the remark this answers is not in the document]
Comment (FrankYFTang) [reply icu4c/source/common/ucurr.cpp]: that will be a behavior change that you ask me not to do in this PR so I take it out and will put that into a later PR.

File: icu4c/source/common/ucurr.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (FrankYFTang) [reply icu4c/source/common/ucurr.cpp]: That will be a behavior change which you ask me not to do in this PR so I won't do here.

File: icu4c/source/common/ucurr.cpp [hunk 6]
```diff
@@ -805,21 +796,19 @@ ucurr_getPluralName(const char16_t* currency,
     rb = ures_getByKey(rb, CURRENCYPLURALS, rb, &ec2);
```
Comment (roubert) [inline icu4c/source/common/ucurr.cpp:796]: What about this one, is there any reason for why it shouldn't use `LocalUResourceBundlePointer` too?

File: icu4c/source/common/ucurr.cpp [hunk 7]
```diff
@@ -805,21 +796,19 @@ ucurr_getPluralName(const char16_t* currency,
...
-        s = ures_getStringByKeyWithFallback(rb, "other", len, &ec2);     
+        s = ures_getStringByKeyWithFallback(curr.getAlias(), "other", len, &ec2);     
         if (U_FAILURE(ec2)) {
-            ures_close(rb);
             // fall back to long name in Currencies
             return ucurr_getName(currency, locale, UCURR_LONG_NAME, 
                                  isChoiceFormat, len, ec);
         }
     }
-    ures_close(rb);
```
Comment (roubert) [inline icu4c/source/common/ucurr.cpp:822]: Is it really correct to now never close `rb`?

File: icu4c/source/common/ucurr.cpp [hunk 8]
```diff
@@ -957,7 +941,7 @@ toUpperCase(const char16_t* source, int32_t len, const char* locale) {
     u_strToUpper(dest, destLen, source, len, locale, &ec);
     if (U_FAILURE(ec)) {
         u_memcpy(dest, source, len);
-    } 
+    }
```
Comment (roubert) [inline icu4c/source/common/ucurr.cpp:944]: Unrelated formatting change.

File: icu4c/source/common/ucurr.cpp [hunk 9]
```diff
@@ -1009,113 +993,80 @@ collectCurrencyNames(const char* locale,
...
-            (*currencyNames)[*total_currency_name_count].flag = NEED_TO_BE_DELETED;
-            (*currencyNames)[(*total_currency_name_count)++].currencyNameLen = len;
+
+            (*currencyNames)[(*total_currency_name_count)++] = {iso, upperName, len, NEED_TO_BE_DELETED};
 
             // put (iso, 3, and iso) in to array
             // Add currency ISO code.
-            (*currencySymbols)[*total_currency_symbol_count].IsoCode = iso;
-            (*currencySymbols)[*total_currency_symbol_count].currencyName = (char16_t*)uprv_malloc(sizeof(char16_t)*3);
+            upperName = (char16_t*)uprv_malloc(sizeof(char16_t)*3);
```
Comment (roubert) [inline icu4c/source/common/ucurr.cpp:1036]: There's nothing to gain by reusing the `upperName` variable, instead allocating a new one, like eg. `char16_t* isoUChars` will make it clearer to the reader what's going on.

File: icu4c/source/common/ucurr.cpp [hunk 10]
```diff
@@ -2295,7 +2234,7 @@ ucurr_countCurrencies(const char* locale,
 {
     int32_t currCount = 0;
 
-    if (ec != nullptr && U_SUCCESS(*ec)) 
+    if (ec != nullptr && U_SUCCESS(*ec))
```
Comment (roubert) [inline icu4c/source/common/ucurr.cpp:2237]: Unrelated formatting change.

File: icu4c/source/common/ucurr.cpp [hunk 11]
```diff
@@ -2320,41 +2259,38 @@ ucurr_countCurrencies(const char* locale,
...
+                    LocalUResourceBundlePointer toRes(ures_getByKey(currencyRes.getAlias(), "to", nullptr, &localStatus));
+                    const int32_t *toArray = ures_getIntVector(toRes.getAlias(), &toLength, &localStatus);
 
                     currDate64 = (int64_t)toArray[0] << 32;
                     currDate64 |= ((int64_t)toArray[1] & (int64_t)INT64_C(0x00000000FFFFFFFF));
                     UDate toDate = (UDate)currDate64;
 
-                    if ((fromDate <= date) && (date < toDate))
-                    {
+                    if ((fromDate <= date) && (date < toDate)) {
```
Comment (roubert) [inline icu4c/source/common/ucurr.cpp:2291]: Unrelated formatting change.

Just to be clear about why I think this matters: When I'm debugging ICU and trying to find where and when some behaviour changed and therefore look through the blame log, if it there on a line of code says that the last change was made by a commit with a description like _"Fix error checking …"_ then that's a change I'll have to take a closer look at and if I then see that the change that commit did to this particular line actually wasn't anything but some unrelated formatting change then that's wasted time, but if a change was made by a commit with a description like _"Fix formatting …"_ then I immediately know that I can skip that one, saving time and effort.

**Review by roubert — CHANGES_REQUESTED** [review]

File: icu4c/source/common/ucurr.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (roubert) [reply icu4c/source/common/ucurr.cpp]: > The current format make it hard to read now.

I wholeheartedly agree, much of the ICU source code is so inconsistently formatted as to make it unnecessarily hard to work with.

> Would you accept a PR only change the format later?

I'd be delighted.

File: icu4c/source/common/ucurr.cpp [hunk 12]
```diff
@@ -2703,19 +2609,18 @@ ucurr_getNumericCode(const char16_t* currency) {
     if (currency && u_strlen(currency) == ISO_CURRENCY_CODE_LENGTH) {
         UErrorCode status = U_ZERO_ERROR;
 
-        UResourceBundle *bundle = ures_openDirect(nullptr, "currencyNumericCodes", &status);
-        ures_getByKey(bundle, "codeMap", bundle, &status);
+        LocalUResourceBundlePointer bundle(ures_openDirect(nullptr, "currencyNumericCodes", &status));
```
Comment (roubert) [inline icu4c/source/common/ucurr.cpp:2612]: This is different from what you're doing in `_findMetaData()` above, there the result from `ures_openDirect()` is stored in a plain pointer upon which `ures_close()` never gets called while the result from the subsequent call to `ures_getByKey()` is stored in a `LocalUResourceBundlePointer`, the opposite of what's being done here. Why are these different, is that intentional, is that correct?

File: icu4c/source/common/ucurr.cpp (hunk 6 above)
Comment (FrankYFTang) [reply icu4c/source/common/ucurr.cpp:796]: I talk with Markus about this. ures_getByKey or ures_getByKeyWithFallback  take the third parameter as an optional fill in, in this case, the third parameter is rb which mean it will put the result into rb without allocate additional memory. so the line of LocalUResourceBundlePointer curr will put that pointer into a LocalUResourceBundlePointer.

File: icu4c/source/common/ucurr.cpp (hunk 7 above)
Comment (FrankYFTang) [reply icu4c/source/common/ucurr.cpp:822]: destrutor of curr will close it

File: icu4c/source/common/ucurr.cpp (hunk 10 above)
Comment (FrankYFTang) [reply icu4c/source/common/ucurr.cpp:2237]: I think we should remove those junk whenever we see it. I do not mind to revert the { change, but this is an extra space or tab.

**Review by richgillam — APPROVED** [review]
LOKTM

**Review by roubert — APPROVED** [review]
LGTM

**Comment by markusicu** [comment]
PS: The internal ures_... functions are another part of ICU that would benefit from modernization. In particular, the fillIn parameters were useful for avoiding some memory allocations, but are hard to get right, and hard to look at. If we did this with modern C++, we would return C++ objects (at least structs with destructors) by value, and move operators...
